Implementation of steering wheel position

Brandnewson opened the pull request:
Had troubles trying to find documentation related to arduino as the github links don't work. But after some digging these were found:

[Arduino code examples
](https://github.com/WITMOTION/WitStandardModbus_WT901C485/blob/main/Arduino/Arduino_sdk/examples/wit_c_sdk_normal/wit_c_sdk_normal.ino)
[Physical installation of gyro on youtube
](https://youtube.com/playlist?list=PL43tdDrVL_VBgiQTMntAm0whkGFbGBdHX)
[Google Drive
](https://drive.google.com/open?id=1dWvJU2Ug7MpcwTPWoARw2KJI8oYbuVEW)

Following changes implemented:
- Changed daq_SteeringWhlPos_t to float32_t to reflect data received
- implemented function to return fAngle as similar to the example code given.

Potential issues:
- As the example returns the data in an array, I'm not sure how the data from the gyro is translated. I can convert it into an array if needed.

Brandnewson commented:
Made a little mistake in submitting a pull request hence the weird history here. But otherwise let me know how I can improve my code :)

Brandnewson commented:
Hi @MellingJJ, is the code still missing? It should be there now.

Brandnewson commented:
Added a REG.h module that was missing earlier. Should be working fine now.

MellingJJ commented:
@Brandnewson I think you've got yourself a little stuck with this one which is fair as there isn't much information in this module, and no readily available Arduino libraries. So I'll help you could a little here. You'll see in my commit that I've added REG and wit_c_sdk.h from the repo of the example you outlined. The example uses these as a library to make handle some of the protocol handling stuff. So with the libraries now in the project, you can use the example.

So start off we have initialization, this should be in gpio_init() function and we should have the following - 
```
        Wire.begin();
        Wire.setClock(400000);
	WitInit(WIT_PROTOCOL_I2C, 0x50);
	WitI2cFuncRegister(IICwriteBytes, IICreadBytes);
	WitRegisterCallBack(CopeSensorData);
  WitDelayMsRegister(Delayms);
	Serial.print("\r\n********************** wit-motion IIC example  ************************\r\n");
	AutoScanSensor();
```
Then for the read, you'll need something like - 
```
        WitReadReg(AX, 12);
        float32_t strWhlAngle_deg = sReg[Roll+2] / 32768.0f * 180.0f;   // Get Z axis angle (this might need to change depending on position of module)
```

Also note you'll have to add the following functions from the example for this to work - 
```
static void AutoScanSensor(void);
static void CopeSensorData(uint32_t uiReg, uint32_t uiRegNum);   
static int32_t IICreadBytes(uint8_t dev, uint8_t reg, uint8_t *data, uint32_t length);
static int32_t IICwriteBytes(uint8_t dev, uint8_t reg, uint8_t* data, uint32_t length);
static void Delayms(uint16_t ucMs);
```

Try talk to Cam about getting HW to test as this might help you get this code working

Brandnewson commented:
hi @MellingJJ. Thanks for the information! I was a little stuck here haha. 

I'll have a look at it later and get it done today. 

Appreciate the patience and guidance!

Brandnewson commented:
Debugging is taking longer than expected so the code will probably only be ready by 5th July Evening. Sorry for the delay.

Brandnewson commented:
Hi @MellingJJ, I've done mostly as described, but faced an issue as described by the summary below.

### gpio.cpp
- included necessary libraries, headers, & linked files
- Added the lines of code as described in github issue here. 
- added calculation for steering wheel position, but will need revising according to module position and the error below


### wit_c_sdk.c
- the file compares signed & unsigned ints. 
Should I change them all to unsigned? 

- Functions from example all added, but faced the following error regarding IICreadBytes & IICwriteBytes: 
**Error: 'class TwoWire' has no member named 'clearWireTimeoutFlag'**

Wire.h library either not updated or is not using the Arduino 'avr' branch which includes a setWireTimeout

**Functions IICreadBytes & IICwriteBytes currently just return 1.**

I'm unsure what the best solution might be for this. Let me know how I can best move forward. :)

Brandnewson commented:
Still trying to find a fix for the wire.h library, but in the meantime I commented out the bits of code that has been giving issues.

MellingJJ commented:
Hi @Brandnewson, thank you for the update.
I don't have my laptop at the moment so I can't properly look into this, but my suggestion would be as you have done, is to remove the wire timeout functions. This should still allow the code to work as necessary, but maybe a little less robust? 

I haven't looked properly at the wit lib stuff so don't know exactly how trying to work but as long as it's only timeout functions that aren't working I think it should be fine to remove them

Brandnewson commented:
@MellingJJ okay got it! let me try that out first.
